here-naksha-lib-ext-manager/src/main/java/com/here/naksha/lib/extmanager/ExtensionCache.java

```diff
@@ -0,0 +1,221 @@
+      ClassLoader loader;
+      try {
+        loader = ClassLoaderHelper.getClassLoader(
+            result.getValue(), extensionConfig.getWhilelistDelegateClass());
+      } catch (Exception e) {
+        logger.error("Failed to load extension jar " + result.getKey().getExtensionId(), e);
+        return;
+      }
+
+      Object object = null;
+      KVPair<Extension, ClassLoader> prevValue = null;
+      if (result.getKey().getInitClassName() == null
+          || result.getKey().getInitClassName().isEmpty()) {
+        prevValue = loaderCache.put(
+            result.getKey().getExtensionId(), new KVPair<Extension, ClassLoader>(result.getKey(), loader));
+      } else {
+        try {
+          Class<?> clz = loader.loadClass(result.getKey().getInitClassName());
+          object = clz.getConstructor(INaksha.class, Object.class)
+              .newInstance(naksha, result.getKey().getProperties());
+          prevValue = loaderCache.put(
+              result.getKey().getExtensionId(),
+              new KVPair<Extension, ClassLoader>(result.getKey(), loader));
+        } catch (ClassNotFoundException
+            | InvocationTargetException
+            | InstantiationException
+            | NoSuchMethodException
+            | IllegalAccessException e) {
+          logger.error(
+              String.format(
+                  "Failed to instantiate class %s for extension %s ",
+                  result.getKey().getInitClassName(),
+                  result.getKey().getExtensionId()),
+              e);
+        }
+      }
+      if (prevValue != null) {
+        ClassLoader prevLoader = prevValue.getValue();
+        prevLoader = null;
+      }
+    }
+  }
+
+  private boolean isLoaderMappingExist(Extension extension) {
+    boolean isEqual = loaderCache.containsKey(extension.getExtensionId());
+    if (isEqual) {
+      Extension exExtension = loaderCache.get(extension.getExtensionId()).getKey();
+      isEqual = exExtension.getUrl().equals(extension.getUrl())
+          && exExtension.getVersion().equals(extension.getVersion())
+          && exExtension.getInitClassName().equals(extension.getInitClassName());
+    }
+    return isEqual;
+  }
+
+  /**
+   * Lamda function which will initiate the downloading for extension jar
+   */
+  private KVPair<Extension, File> downloadJar(Extension extension) {
+    logger.info("Downloading jar {} with version {} ", extension.getExtensionId(), extension.getVersion());
+    FileClient client = getJarClient(extension.getUrl());
+    File file = null;
+    try {
+      file = client.getFile(extension.getUrl());
+    } catch (IOException | SdkClientException e) {
+      logger.error("Failed to fetch jar {} ", extension.getUrl());
+    }
+    return new KVPair<Extension, File>(extension, file);
+  }
+
+  // TODO: Can be moved to factory function. Since not used elsewhere placed it inside this class
+  protected FileClient getJarClient(String url) {
```

Comment:
From what I can say, this function is not thread safe, but possibly used from multiple threads in parallel? You can simply make it synchronized I would say as an easy fix, seems like it is only used when downloading the JAR and the time for the mutex does not really effect the total download time, but you should not allow accessing a HashMap (which is not thread safe) from multiple threads (unless I miss some synchronization point).

Reply:
It may be worth considering population of this map during constructor itself and I would also make it immutable. As from what I understand, it seems to be used as a way to return singleton factory instance, so is not expected to change once populated.

Reply:
Map is static final now. And has been initialised within static block. Since as of now it only contains 2 types of FileClient(s3 or Filesystem)